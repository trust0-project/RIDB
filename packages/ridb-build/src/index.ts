


import { defineConfig, Format } from 'tsup';
import { NodeResolvePlugin } from '@esbuild-plugins/node-resolve';
import fs from 'fs';
import path from 'path';

const packagesDir = path.resolve(process.cwd(), "../../", "packages/ridb-core");
export const wasmPlugin = {
    name: 'wasm',
    setup(build:any) {
        build.onResolve({ filter: /\.wasm$/ }, (args:any) => {
            if (fs.existsSync(path.resolve(packagesDir, args.path))) {
                return { path: path.resolve(packagesDir, args.path), namespace: 'wasm' };
            }
            return { path: path.resolve('../../node_modules', args.path), namespace: 'wasm' };
        });
        build.onLoad({ filter: /.*/, namespace: 'wasm' }, async (args:any) => {
            const buffer = await fs.promises.readFile(args.path);
            const base64 = buffer.toString('base64');
            return {
                contents: `export default Buffer.from("${base64}", "base64")`,
                loader: 'js',
            };
        });
    },
};

export const plugins = [
    NodeResolvePlugin({
        extensions: ['.ts', '.js', '.wasm'],
        onResolved: (resolved) => {
            if (resolved.includes('node_modules')) {
                return {
                    external: true,
                }
            }
            return resolved
        },
    }),
]

export default function createConfig({format, entry, banner, platform}:{
  format: Format | Format[] | undefined,
  entry: string[] | undefined,
  banner?: {js:string},
  platform?: 'neutral' | 'node' | 'browser'
}) {
  return defineConfig(({ watch }) => ({
    entry,
    format,
    outDir: 'build',
    target: 'esnext',
    minify: false,
    clean: false,
    esbuildPlugins: [
      wasmPlugin,
      ...plugins
    ],
    banner,
    esbuildOptions:(options, context) => {
      options.platform = platform || 'neutral'
    },
    external: [
      'buffer',
      'next',
      'vitest', 
      'react-server-dom-webpack', 
      'tsup', 
      'react-server-dom-webpack/client.edge',
      '@trust0/ridb/worker'
    ],
  }));
}