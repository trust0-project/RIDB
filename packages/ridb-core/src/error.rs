use serde::{Deserialize, Serialize};
use wasm_bindgen::prelude::*;
use crate::utils::extract_property;

#[wasm_bindgen]
#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum Errors {
    Error,
    HookError,
    QueryError,
    SerializationError,
    ValidationError,
    AuthenticationError,
}

#[wasm_bindgen(inspectable)]
#[derive(Serialize, Deserialize, Debug)]
pub struct RIDBError {
    pub(crate) err_type: String,
    pub(crate) message: String,
    pub(crate) code: u32
}

#[wasm_bindgen]
impl RIDBError {

    #[wasm_bindgen(constructor)]
    pub fn new(
        err_type: String,
        message: String,
        code: u32
    ) -> RIDBError {
        RIDBError {
            err_type,
            message,
            code
        }
    }

    #[wasm_bindgen(getter, js_name=type)]
    pub fn get_type(&self) -> String {
        self.err_type.to_string()
    }

    #[wasm_bindgen(getter, js_name=code)]
    pub fn get_code(&self) -> JsValue {
        JsValue::from(
            self.code
        )
    }

    #[wasm_bindgen(getter, js_name=message)]
    pub fn get_message(&self) -> String {
        self.message.to_string()
    }

    #[wasm_bindgen(js_name=from)]
    pub fn get_from(err: JsValue) -> RIDBError {
        RIDBError::from(err)
    }

    fn create_error(message: &str, code: u32, err_type: Errors) -> RIDBError {
        let message_type = match err_type {
            Errors::QueryError => "Query Error".to_string(),
            Errors::AuthenticationError => "Authentication Error".to_string(),
            Errors::SerializationError => "Serialization Error".to_string(),
            Errors::ValidationError => "Validation Error".to_string(),
            Errors::HookError => "Hook Error".to_string(),
            _ => "Error".to_string(),
        };
        let complete_message=  format!("{}: {}", message_type, message);
        RIDBError {
            err_type: format!("{:?}", err_type),
            code,
            message: complete_message.as_str().to_string()
        }
    }

    #[wasm_bindgen]
    pub fn error(err: &str, code: u32) -> RIDBError {
        RIDBError::create_error(err, code, Errors::Error)
    }

    #[wasm_bindgen]
    pub fn query(err: &str, code: u32) -> RIDBError {
        RIDBError::create_error(err, code, Errors::QueryError)
    }

    #[wasm_bindgen]
    pub fn authentication(err: &str, code: u32) -> RIDBError {
        RIDBError::create_error(err, code, Errors::AuthenticationError)
    }

    #[wasm_bindgen]
    pub fn serialisation(err: &str, code: u32) -> RIDBError {
        RIDBError::create_error(err, code, Errors::SerializationError)
    }

    #[wasm_bindgen]
    pub fn validation(err: &str, code: u32) -> RIDBError {
        RIDBError::create_error(err, code, Errors::ValidationError)
    }

    #[wasm_bindgen]
    pub fn hook(err: &str, code: u32) -> RIDBError {
        RIDBError::create_error(err, code, Errors::HookError)
    }
}

impl From<serde_wasm_bindgen::Error> for RIDBError {
    fn from(error: serde_wasm_bindgen::Error) -> RIDBError {
        RIDBError::serialisation(
            error.to_string().replace("Error: ", "").as_str(),
            0
        )
    }
}

impl From<&String> for RIDBError {
    fn from(error:&String) -> RIDBError {
        RIDBError::error(
            error.as_str(),
            0
        )
    }
}

impl From<&str> for RIDBError {
    fn from(error:&str) -> RIDBError {
        RIDBError::error(
            error,
            0
        )
    }
}

impl From<String> for RIDBError {
    fn from(error:String) -> RIDBError {
        RIDBError::error(
            error.as_str(),
            0
        )
    }
}

impl From<JsValue> for RIDBError {
    fn from(error: JsValue) -> RIDBError {
        let err_type_js = extract_property::<String>(&error, "type").unwrap_or("Error".to_string());
        let code = extract_property::<u32>(&error, "code").unwrap_or(0);
        let message = extract_property::<String>(&error, "message").expect("Invalid JS Error no message is available");
        RIDBError {
            err_type:err_type_js.to_string(),
            code,
            message: message.as_str().to_string()
        }
    }
}
