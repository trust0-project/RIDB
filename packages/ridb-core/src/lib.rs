use wasm_bindgen::prelude::*;
use wasm_bindgen::JsValue;
use std::panic;

pub mod error;
pub mod utils;
pub mod schema;
pub mod collection;
pub mod storages;
pub mod storage;
pub mod database;
pub mod query;
pub mod operation;
pub mod plugin;

#[wasm_bindgen(start)]
pub fn main_js() -> Result<(), JsValue> {
    panic::set_hook(Box::new(console_error_panic_hook::hook));
    #[cfg(feature = "console_error_panic_hook")]
    console_error_panic_hook::set_once();
    Ok(())
}

#[wasm_bindgen]
pub fn is_debug_mode() -> bool {
    get_debug_mode()
}

fn get_debug_mode() -> bool {
    use wasm_bindgen::prelude::*;
    use js_sys::Reflect;

    if let Some(win) = web_sys::window() {
        // Browser environment
        win.local_storage()
            .ok()
            .flatten()
            .and_then(|storage| storage.get_item("DEBUG").ok().flatten())
            .map(|debug_str| {
                debug_str
                    .split(',')
                    .any(|s| s == "ridb" || s.starts_with("ridb:*"))
            })
            .unwrap_or(false)
    } else {
        // Node.js environment
        // Access process.env.DEBUG directly
        let global = js_sys::global();

        let process = Reflect::get(&global, &JsValue::from_str("process")).ok();
        let env = process
            .as_ref()
            .and_then(|proc| Reflect::get(proc, &JsValue::from_str("env")).ok());
        let debug_var = env
            .as_ref()
            .and_then(|env| Reflect::get(env, &JsValue::from_str("DEBUG")).ok());

        if let Some(debug_js_value) = debug_var {
            if let Some(debug_str) = debug_js_value.as_string() {
                debug_str
                    .split(',')
                    .any(|s| s == "ridb" || s.starts_with("ridb:*"))
            } else {
                false
            }
        } else {
            false
        }
    }
}

mod logger {
    use wasm_bindgen::prelude::*;
    use web_sys::console;

    pub struct Logger;

    impl Logger {
        pub fn error(component: &str, message: &JsValue) {
            Logger::err_1(component, message);
        }
        pub fn log(component: &str, message: &JsValue) {
            Logger::log_1(component, message);
        }
        pub fn debug(component: &str, message: &JsValue) {
            if crate::is_debug_mode() {
                Logger::log_1(component, message);
            }
        }

        fn log_1(component: &str, message: &JsValue) {
            console::log_1(
                &JsValue::from(
                    format!("[{}] {:?}", component, message)
                )
            );
        }

        fn err_1(component: &str, message: &JsValue) {
            console::log_1(
                &JsValue::from(
                    format!("[{}] {:?}", component, message)
                )
            );
        }
    }
}