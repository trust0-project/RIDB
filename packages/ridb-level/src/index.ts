/**
 * @packageDocumentation
 *
 * <p align="center">
 *  <img src="https://cdn.jsdelivr.net/gh/trust0-project/ridb@latest/docs/logo.svg" alt="JavaScript Database" />
 *  <br />
 *  <br />
 *  <h3 align="center">A secure light-weight and dependency free database wrapper for the web.</h3>
 * </p>
 * <p align="center">
 *   <a href="https://github.com/trust0-project/RIDB/releases"><img src="https://img.shields.io/github/v/release/trust0-project/ridb?color=%23ff00a0&include_prereleases&label=version&sort=semver&style=flat-square"></a>
 *   <a href="#"><img src="https://img.shields.io/npm/types/rxdb?style=flat-square"></a>
 *   <a href="https://raw.githubusercontent.com/trust0-project/RIDB/refs/heads/main/LICENSE"><img src="https://img.shields.io/github/license/trust0-project/ridb?style=flat-square"></a>
 *   <a href="https://www.npmjs.com/package/@trust0/ridb"><img src="https://img.shields.io/npm/dm/@trust0/ridb?color=c63a3b&style=flat-square"></a>   
 * </p>
 * # Features
 * Will create a levelDB folder in the current working directory with the db name.
 * Specifying a different path is not yet supported.
 * 
 * # SDK Rerefence
 */
import path from 'path';
import { 
    QueryOptions, 
    SchemaTypeRecord, 
    Doc, 
    Operation,
    QueryType,    
} from "@trust0/ridb-core";

import {WasmInternal} from "@trust0/ridb";

import type { ClassicLevel } from "classic-level";
type Level = ClassicLevel<string, string>;




const { BaseStorage, Query, OpType } = WasmInternal as typeof import("@trust0/ridb-core");

export class LevelDB<T extends SchemaTypeRecord> extends BaseStorage<T> {

    static async create<SchemasCreate extends SchemaTypeRecord>(
        name: string,
        schemas: SchemasCreate,
        options: any
    ): Promise<LevelDB<SchemasCreate>> {
        const {ClassicLevel} = await import("classic-level");
        const level = new ClassicLevel(path.resolve(process.cwd(), "./.db/" + name));
        const db = new LevelDB<SchemasCreate>(level, name, schemas, options);
        return db;
    }
    constructor(private db: Level, name: string, schemas: T, options: any) {
        super(name, schemas, options);
    }
    /** Start the database */
    async start(): Promise<void> {
        await this.db.open();
    }
    /** Close the database */
    async close(): Promise<void> {
        await this.db.close(); // Close the database
    }
    /** Find a document by its ID */
    async findDocumentById(
        collectionName: keyof T,
        id: string
    ): Promise<Doc<T[keyof T]> | null> {
        const key = `${String(collectionName)}:${id}`;
        try {
            const value = await this.db.get(key);
            if (!value) {
                return null
            }
            const doc = JSON.parse(value);
            return doc;
        } catch (err: any) {
            if (err.notFound) {
                return null;
            } else {
                throw err;
            }
        }
    }
    /** Write an operation (insert, update, delete) */
    async write(op: Operation<T[keyof T]>): Promise<Doc<T[keyof T]>> {
        const collectionName = String(op.collection);
        const primaryKey = this.getSchema(collectionName).primaryKey;

        switch (op.opType) {
            case OpType.CREATE: {
                const id = primaryKey  in op.data ? (op.data as any)[primaryKey] : null;
                if (!id) {
                    throw new Error("Document ID is required");
                }
                const key = `${collectionName}:${id}`;
                const value = JSON.stringify(op.data);
                await this.db.put(key, value);
                break;
            }
            case OpType.UPDATE: {
                const id = primaryKey  in op.data ? (op.data as any)[primaryKey] : null;
                if (!id) {
                    throw new Error("Document ID is required");
                }
                const existingRecord = await this.findDocumentById(collectionName, id);
                if (!existingRecord) {
                    throw new Error("Document ID not found");
                }
                const key = `${collectionName}:${id}`;
                const value = JSON.stringify({
                    ...existingRecord,
                    ...op.data
                });
                await this.db.put(key, value);
                break;
            }
            case OpType.DELETE: {
                const key = `${collectionName}:${op.data}`;
                await this.db.del(key);
                break;
            }
            default:
                throw new Error(`Unknown operation type: ${op.opType}`);
        }
        return op.data;
    }
    /** Count documents matching a query (supports offset & limit) */
    async count(
        collectionName: keyof T,
        query: QueryType<T[keyof T]>,
        options?: QueryOptions
    ): Promise<number> {
        const collectionPrefix = `${String(collectionName)}:`;
        const schema = this.getSchema(String(collectionName));
        const queryInstance = new Query(query, schema);

        // Retrieve limit and offset from options, using defaults if undefined
        const limit = options?.limit ?? -1;  // -1 indicates "no limit" for classic-level
        const offset = options?.offset ?? 0; // default to 0 (no offset)

        let matchedCount = 0;
        let skipCount = offset;

        // If a limit is set, set maxRead = limit + offset
        // Otherwise, let maxRead be -1 (no limit)
        const maxRead = limit > 0 ? offset + limit : -1;

        for await (const [key, value] of this.db.iterator({
            gte: collectionPrefix,
            lt: `${collectionPrefix}\xFF`,
            limit: maxRead
        })) {
            const doc = JSON.parse(value);

            if (!this.core.matchesQuery(doc, queryInstance)) {
                continue;
            }

            // Skip documents until offset is reached
            if (skipCount > 0) {
                skipCount--;
                continue;
            }

            matchedCount++;

            // If we've reached the limit, break early
            if (limit > 0 && matchedCount >= limit) {
                break;
            }
        }

        return matchedCount;
    }
    /** Find documents matching a query with pagination */
    async find(
        collectionName: keyof T,
        query: QueryType<T[keyof T]>,
        options?: QueryOptions
    ): Promise<Doc<T[keyof T]>[]> {
        const collectionPrefix = `${String(collectionName)}:`;
        const schema = this.getSchema(String(collectionName));
        const queryInstance = new Query(query, schema);

        // Retrieve limit and offset from the options, using defaults if undefined
        const limit = options?.limit ?? -1;  // -1 indicates "no limit" for classic-level
        const offset = options?.offset ?? 0; // default to 0 (no offset)

        const docs: Doc<T[keyof T]>[] = [];
        let skipCount = offset;

        // If a limit is set, set maxRead = limit + offset
        // Else let maxRead be -1 (no limit) so we read all
        const maxRead = limit > 0 ? offset + limit : -1;

        for await (const [key, value] of this.db.iterator({
            gte: collectionPrefix,
            lt: `${collectionPrefix}\xFF`,
            limit: maxRead
        })) {
            const doc = JSON.parse(value);
            if (!this.core.matchesQuery(doc, queryInstance)) {
                continue;
            }

            // Skip documents until offset is reached
            if (skipCount > 0) {
                skipCount--;
                continue;
            }

            docs.push(doc);

            // If we've reached the limit, break early
            if (limit > 0 && docs.length >= limit) {
                break;
            }
        }

        return docs;
    }
}
