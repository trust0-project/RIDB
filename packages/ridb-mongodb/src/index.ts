/**
 * @packageDocumentation
 *
 * <p align="center">
 *      <img src="../../../resources/ridb-dark.svg" alt="JavaScript Database" />
 *  <br />
 *  <br />
 *  <h3 align="center">A secure light-weight and dependency free database wrapper for the web.</h3>
 * </p>
 * <p align="center">
 *   <a href="https://github.com/trust0-project/RIDB/releases"><img src="https://img.shields.io/github/v/release/trust0-project/ridb?color=%23ff00a0&include_prereleases&label=version&sort=semver&style=flat-square"></a>
 *   <a href="#"><img src="https://img.shields.io/npm/types/rxdb?style=flat-square"></a>
 *   <a href="https://raw.githubusercontent.com/trust0-project/RIDB/refs/heads/main/LICENSE"><img src="https://img.shields.io/github/license/trust0-project/ridb?style=flat-square"></a>
 *   <a href="https://www.npmjs.com/package/@trust0/ridb"><img src="https://img.shields.io/npm/dm/@trust0/ridb?color=c63a3b&style=flat-square"></a>
 * </p>
 *
 * # Features
 * Provides MongoDB storage backend for RIDB with full CRUD operations and query support.
 * Supports all MongoDB connection options and authentication methods.
 *
 * # SDK Rerefence
 */
/** biome-ignore-all lint/suspicious/noExplicitAny: Not needed here */

import { WasmInternal } from "@trust0/ridb";
import {
  type BaseStorageOptions,
  type BaseStorage as BaseStorageType,
  type Doc,
  type Operation,
  OpType,
  type QueryOptions,
  type QueryType,
  type SchemaTypeRecord,
} from "@trust0/ridb-core";

import type { Collection, Db, MongoClient } from "mongodb";

export interface MongoDBConfig {
  /** MongoDB connection URL */
  url?: string;
  /** Database name in MongoDB */
  dbName?: string;
  /** MongoDB client options */
  mongoOptions?: Record<string, any>;
}

export type MongoDBStorageOptions = BaseStorageOptions & MongoDBConfig;

/**
 * Create a MongoDB storage instance
 * @public
 * @returns A factory function that creates MongoDB storage instances
 */
export async function createMongoDB<T extends SchemaTypeRecord>(): Promise<typeof BaseStorageType<T>> {
  const { BaseStorage } = await WasmInternal();

  // We need to extend the actual BaseStorage from WasmInternal
  /**
   * MongoDB storage implementation class
   * @public
   */
  class MongoDBStorage<T extends SchemaTypeRecord> extends BaseStorage<T> {
    private client?: MongoClient;
    private db?: Db;
    private mongoConfig: MongoDBConfig;

    constructor(name: string, schemas: T, options: MongoDBStorageOptions = {} as MongoDBStorageOptions) {
      super(name, schemas, options);
      this.mongoConfig = options;
    }

    /**
     * Utility method to recursively convert ObjectId values to strings throughout an object
     * This ensures that MongoDB's ObjectId instances are never returned to the application layer,
     * maintaining consistency with RIDB's string-based approach and preventing serialization issues.
     * Also removes MongoDB's automatic _id field entirely from results.
     * @private
     */
    private convertObjectIdsToStrings(obj: any): any {
      if (!obj || typeof obj !== "object") {
        return obj;
      }

      // Handle ObjectId
      if (obj.constructor?.name === "ObjectId" || (obj._bsontype && obj._bsontype === "ObjectId")) {
        const stringValue = obj.toString();
        return stringValue;
      }

      // Handle arrays
      if (Array.isArray(obj)) {
        return obj.map((item) => this.convertObjectIdsToStrings(item));
      }

      // Handle plain objects
      const converted: any = {};
      for (const [key, value] of Object.entries(obj)) {
        // Skip MongoDB's _id field entirely
        if (key === "_id") {
          continue;
        }
        converted[key] = this.convertObjectIdsToStrings(value);
      }

      return converted;
    }

    /**
     * Create a new MongoDB storage instance
     * @param name - Database name
     * @param schemas - Collection schemas
     * @param options - Storage options including MongoDB connection details
     * @returns A new Instance of MongoDB storage
     * @public
     */
    static async create<SchemasCreate extends SchemaTypeRecord>(name: string, schemas: SchemasCreate, options: MongoDBStorageOptions = {}) {
      const db = new MongoDBStorage<SchemasCreate>(name, schemas, options);
      return db;
    }

    /** Start the database connection */
    async start(): Promise<void> {
      const { MongoClient } = await import("mongodb");
      const url = this.mongoConfig.url || process.env.MONGODB_URL;
      if (!url) {
        throw new Error("MONGODB_URL is required");
      }
      const dbName = this.mongoConfig.dbName || (this as any).name;
      this.client = new MongoClient(url, {
        connectTimeoutMS: 5000,
        serverSelectionTimeoutMS: 5000,
        socketTimeoutMS: 5000,
        maxPoolSize: 10,
        minPoolSize: 5,
        maxIdleTimeMS: 5000,
        maxConnecting: 5,
      });
      await this.client.connect();
      this.db = this.client.db(dbName);
    }

    /** Close the database connection */
    async close(): Promise<void> {
      if (this.client) {
        await this.client.close();
        this.client = undefined;
        this.db = undefined;
      }
    }

    /** Get MongoDB collection for a given collection name */
    private getCollection(collectionName: string): Collection {
      if (!this.db) {
        throw new Error("Database not started. Call start() first.");
      }
      const collection = this.db.collection(collectionName);
      return collection;
    }

    /** Find a document by its ID */
    async findDocumentById(collectionName: keyof T, id: string): Promise<Doc<T[keyof T]> | null> {
      const collection = this.getCollection(String(collectionName));
      const schema = this.getSchema(String(collectionName));
      const primaryKey = schema.primaryKey;

      const query = { [primaryKey]: id };

      const doc = await collection.findOne(query);

      if (!doc) {
        return null;
      }

      const converted = this.convertObjectIdsToStrings(doc) as Doc<T[keyof T]>;
      return converted;
    }

    /** Write an operation (insert, update, delete) */
    async write(op: Operation<T[keyof T]>): Promise<Doc<T[keyof T]>> {
      const collectionName = String(op.collection);
      const collection = this.getCollection(collectionName);
      const { primaryKey } = this.getSchema(collectionName);

      switch (op.opType) {
        case OpType.CREATE: {
          const id = primaryKey in op.data ? (op.data as any)[primaryKey] : null;

          if (!id) {
            throw new Error("Document ID is required");
          }

          // Check if document already exists
          const existQuery = { [primaryKey]: id };
          const existing = await collection.findOne(existQuery);

          if (existing) {
            throw new Error(`Document with ${primaryKey} '${id}' already exists`);
          }

          // Prevent MongoDB from auto-generating _id by explicitly setting it to null
          const docToInsert = { ...(op.data as any) };

          await collection.insertOne(docToInsert);

          const result = this.convertObjectIdsToStrings(op.data);
          return result;
        }
        case OpType.UPDATE: {
          const id = primaryKey in op.data ? (op.data as any)[primaryKey] : null;

          if (!id) {
            throw new Error("Document ID is required");
          }

          const updateQuery = { [primaryKey]: id };
          const updateData = { $set: op.data };

          const result = await collection.updateOne(updateQuery, updateData);

          if (result.matchedCount === 0) {
            throw new Error("Document ID not found");
          }

          const convertedResult = this.convertObjectIdsToStrings(op.data);
          return convertedResult;
        }
        case OpType.DELETE: {
          const id = op.data;

          const deleteQuery = { [primaryKey]: id };

          const result = await collection.deleteOne(deleteQuery);

          if (result.deletedCount === 0) {
            throw new Error("Document ID not found");
          }

          const convertedResult = this.convertObjectIdsToStrings(op.data);
          return convertedResult;
        }
        default:
          throw new Error(`Unknown operation type: ${op.opType}`);
      }
    }

    /** Convert RIDB query to MongoDB filter */
    private convertQueryToMongoFilter(query: QueryType<T[keyof T]>, _schema: any): any {
      if (!query || typeof query !== "object") {
        return {};
      }

      const filter: any = {};

      for (const [key, value] of Object.entries(query)) {
        if (value === null || value === undefined) {
          filter[key] = { $exists: false };
        } else if (typeof value === "object" && value !== null) {
          // Convert any ObjectId values to strings in query operators
          if ("$in" in value) {
            filter[key] = { $in: this.convertObjectIdsToStrings(value.$in) };
          } else if ("$nin" in value) {
            filter[key] = { $nin: this.convertObjectIdsToStrings(value.$nin) };
          } else if ("$gt" in value) {
            filter[key] = { $gt: this.convertObjectIdsToStrings(value.$gt) };
          } else if ("$gte" in value) {
            filter[key] = { $gte: this.convertObjectIdsToStrings(value.$gte) };
          } else if ("$lt" in value) {
            filter[key] = { $lt: this.convertObjectIdsToStrings(value.$lt) };
          } else if ("$lte" in value) {
            filter[key] = { $lte: this.convertObjectIdsToStrings(value.$lte) };
          } else if ("$ne" in value) {
            filter[key] = { $ne: this.convertObjectIdsToStrings(value.$ne) };
          } else if ("$regex" in value) {
            filter[key] = { $regex: value.$regex };
          } else {
            // For complex objects, convert ObjectIds and use exact match
            filter[key] = this.convertObjectIdsToStrings(value);
          }
        } else {
          filter[key] = this.convertObjectIdsToStrings(value);
        }
      }

      return filter;
    }

    /** Count documents matching a query (supports offset & limit) */
    async count(collectionName: keyof T, query: QueryType<T[keyof T]>, options?: QueryOptions): Promise<number> {
      const collection = this.getCollection(String(collectionName));
      const schema = this.getSchema(String(collectionName));

      // For MongoDB, we use the native query system instead of the RIDB Query class
      // to get better performance, but we could also use Query class for consistency
      const filter = this.convertQueryToMongoFilter(query, schema);

      const countOptions: any = {};
      if (options?.offset) {
        countOptions.skip = options.offset;
      }
      if (options?.limit) {
        countOptions.limit = options.limit;
      }

      const count = await collection.countDocuments(filter, Object.keys(countOptions).length > 0 ? countOptions : undefined);

      return count;
    }

    /** Find documents matching a query with pagination */
    async find(collectionName: keyof T, query: QueryType<T[keyof T]>, options?: QueryOptions): Promise<Doc<T[keyof T]>[]> {
      const collection = this.getCollection(String(collectionName));
      const schema = this.getSchema(String(collectionName));

      // For MongoDB, we use the native query system for better performance
      const filter = this.convertQueryToMongoFilter(query, schema);

      let findQuery = collection.find(filter);

      // Apply offset and limit if provided
      if (options?.offset) {
        findQuery = findQuery.skip(options.offset);
      }

      if (options?.limit) {
        findQuery = findQuery.limit(options.limit);
      }

      const docs = await findQuery.toArray();

      // Convert ObjectIds to strings and remove _id fields from all documents
      const convertedDocs = docs.map((doc) => this.convertObjectIdsToStrings(doc)) as Doc<T[keyof T]>[];

      return convertedDocs;
    }
  }

  // Return the static create method as the constructor
  return MongoDBStorage as unknown as typeof BaseStorage<T>;
}
