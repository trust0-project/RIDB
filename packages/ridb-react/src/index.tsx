import React, { createContext, useMemo, useContext, useEffect } from 'react';
import { RIDB, SchemaTypeRecord, BasePlugin, MigrationsParameter, WasmInternal } from "@trust0/ridb";
await     WasmInternal();


type DatabaseProps<T extends SchemaTypeRecord> = {
  dbName: string;
  schemas: T;
  plugins?: Array<typeof BasePlugin>;
} & MigrationsParameter<T>;

const DatabaseContext = createContext<RIDB<any> | null>(null);

export type DatabaseComponentProps<T extends SchemaTypeRecord> = DatabaseProps<T> & {
  children?: React.ReactNode;
};

export function useDatabase<T extends SchemaTypeRecord>(): RIDB<T> {
  const context = useContext(DatabaseContext);
  if (!context) {
    throw new Error('useDatabase must be used within a Database provider');
  }
  useEffect(() => {
    WasmInternal()
  }, [WasmInternal]);
  return context as RIDB<T>;
}

export function Database<T extends SchemaTypeRecord>({ children, ...props }: DatabaseComponentProps<T>) {
  const dbInit = props as DatabaseProps<T>;
  const db = useMemo(() => new RIDB<T>(dbInit), [props]);
  return (
    <DatabaseContext.Provider value={db}>
      {children}
    </DatabaseContext.Provider>
  );
}


