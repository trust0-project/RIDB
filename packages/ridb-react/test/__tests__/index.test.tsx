import React from 'react'
import { render, waitFor, screen } from '@testing-library/react'
import { SchemaFieldType, StorageType } from '@trust0/ridb'
import { describe, expect, it, vi } from 'vitest'
import { RIDBDatabase, useRIDB } from '../../src'

const users = {
  version: 0 as const,
  primaryKey: 'id',
  type: SchemaFieldType.object,
  properties: {
      id: {
          type: SchemaFieldType.string,
          maxLength: 60
      },
      name: {
          type: SchemaFieldType.string
      },
      age: {
          type: SchemaFieldType.number
      }
  }
} as const

const schemas = {
  users: users
}
  
type DatabaseSchemas = typeof schemas;

const MyComponent: React.FC = () => {
    const { db, isLoading, error } = useRIDB<DatabaseSchemas>();
    const [isDbReady, setIsDbReady] = React.useState(false);

    React.useEffect(() => {
        const startDb = async () => {
            if (db) {
                await db.start();
                setIsDbReady(true);
            }
        };
        
        if (!isLoading && db) {
            startDb();
        }
    }, [db, isLoading]);

    if (error) {
        return <div data-testid="error">Error: {error.message}</div>;
    }

    if (isLoading) {
        return <div data-testid="loading">Database is initializing...</div>;
    }

    if (!db) {
        return <div data-testid="not-available">Database is not available</div>;
    }

    if (!isDbReady) {
        return <div data-testid="starting">Loading...</div>;
    }

    return (
        <div data-testid="ready-component"> <h1>My Component</h1> </div>
    );
};

// Component that simulates database operations
const MockDataComponent: React.FC = () => {
    const { db, isLoading, error } = useRIDB<DatabaseSchemas>();
    const [status, setStatus] = React.useState<string>('idle');

    React.useEffect(() => {
        const simulateDbOperations = async () => {
            if (db) {
                try {
                    await db.start();
                    setStatus('success');
                } catch (err) {
                    setStatus('error');
                }
            }
        };
        
        if (!isLoading && db) {
            setStatus('processing');
            simulateDbOperations();
        }
    }, [db, isLoading]);

    if (error) {
        return <div data-testid="error">Error: {error.message}</div>;
    }

    if (isLoading) {
        return <div data-testid="loading">Loading database...</div>;
    }

    if (status === 'idle') {
        return <div data-testid="idle">Waiting to start...</div>;
    }

    if (status === 'processing') {
        return <div data-testid="processing">Processing operations...</div>;
    }

    if (status === 'error') {
        return <div data-testid="op-error">Operation failed</div>;
    }

    return (
        <div data-testid="data-ready">
            <h1>Data Operations Complete</h1>
            <div data-testid="user-data">John Doe (Age: 30)</div>
        </div>
    );
};

describe('RIDB React Components', () => {
  // Basic rendering tests
  describe('MyComponent', () => {
    it('renders loading state initially', async () => {
      const { getByTestId, container } = render(
        <RIDBDatabase 
          dbName="testDB" 
          schemas={schemas} 
          startOptions={{
            storageType: StorageType.InMemory, 
            password: "testtrust0"
          }}
        >
          <MyComponent />
        </RIDBDatabase>
      );
      
      expect(getByTestId('loading')).toBeTruthy();
      expect(container).toMatchSnapshot('MyComponent-loading-state');
    });

    it('renders successfully when database is ready', async () => {
      const { container } = render(
        <RIDBDatabase 
          dbName="testDB" 
          schemas={schemas} 
          startOptions={{
            storageType: StorageType.InMemory, 
            password: "testtrust0"
          }}
        >
          <MyComponent />
        </RIDBDatabase>
      );
      
      // Wait for the component to fully load
      await waitFor(() => {
        expect(screen.getByTestId('ready-component')).toBeTruthy();
      }, { timeout: 3000 });
      
      expect(container).toMatchSnapshot('MyComponent-ready-state');
    });
  });
  
  // Data operations tests (using mock)
  describe('Data Operations', () => {
    it('shows correct states during data operations', async () => {
      const { container } = render(
        <RIDBDatabase 
          dbName="testDB" 
          schemas={schemas} 
          startOptions={{
            storageType: StorageType.InMemory, 
            password: "testtrust0"
          }}
        >
          <MockDataComponent />
        </RIDBDatabase>
      );
      
      // First should show loading
      expect(screen.getByTestId('loading')).toBeTruthy();
      expect(container).toMatchSnapshot('MockDataComponent-loading-state');
      
      // Eventually should show data is ready
      await waitFor(() => {
        expect(screen.getByTestId('data-ready')).toBeTruthy();
      }, { timeout: 3000 });
      
      // Should display mocked user data
      expect(screen.getByTestId('user-data').textContent).toContain('John Doe');
      expect(container).toMatchSnapshot('MockDataComponent-data-ready-state');
    });
  });
  
})

