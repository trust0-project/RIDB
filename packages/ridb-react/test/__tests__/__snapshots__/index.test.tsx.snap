// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyComponent > renders correctly when db is ready 1`] = `
<DocumentFragment>
  <div>
     
    <h1>
      My Component
    </h1>
     
  </div>
</DocumentFragment>
`;

exports[`MyComponent > renders correctly while db is loading 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
